test/api/test_arrow.cpp

```diff
@@ -57,20 +61,41 @@ struct MyArrowArrayStream {
 		return my_stream->last_error.c_str();
 	}
 
-	unique_ptr<QueryResult> duckdb_result;
+	QueryResult *duckdb_result;
 	ArrowArrayStream stream;
 	string last_error;
 };
 
-static void test_arrow_round_trip(string q) {
+class MyPythonTableArrowArrayStreamFactory {
+public:
+	explicit MyPythonTableArrowArrayStreamFactory(unique_ptr<QueryResult> duckdb_result)
+	    : arrow_table(move(duckdb_result)) {};
+	static ArrowArrayStream *Produce(uintptr_t factory_ptr) {
+		MyPythonTableArrowArrayStreamFactory *factory = (MyPythonTableArrowArrayStreamFactory *)factory_ptr;
+		if (!factory->stream) {
+			auto table_stream = new MyArrowArrayStream(factory->arrow_table.get());
+			factory->stream = &table_stream->stream;
+		}
+		return factory->stream;
+	};
+	unique_ptr<QueryResult> arrow_table;
+	ArrowArrayStream *stream = nullptr;
+};
+
+static void TestArrowRoundTrip(string q) {
 	DuckDB db(nullptr);
 	Connection con(db);
 
 	// query that creates a bunch of values across the types
 	auto result = con.Query(q);
 	REQUIRE(result->success);
-	auto my_stream = new MyArrowArrayStream(move(result));
-	auto result2 = con.TableFunction("arrow_scan", {Value::POINTER((uintptr_t)&my_stream->stream)})->Execute();
+	unique_ptr<MyPythonTableArrowArrayStreamFactory> stream_factory =
+	    make_unique<MyPythonTableArrowArrayStreamFactory>(move(result));
+	ArrowArrayStream *(*stream_factory_produce)(uintptr_t factory) = MyPythonTableArrowArrayStreamFactory::Produce;
+
+	auto result2 = con.TableFunction("arrow_scan", {Value::POINTER((uintptr_t)stream_factory.get()),
+	                                                Value::POINTER((uintptr_t)stream_factory_produce)})
```

Comment:
Maybe I'm missing something, but isn't `stream_factory_produce` a static function pointer? Can't you call `MyPythonTableArrowArrayStreamFactory::Produce` directly from the factory, or even just add this pointer to the factory, rather than adding another argument to the function in SQL land?

Reply:
The problem with that is that the factory must hold a py::object arrow_table;
To call it directly from the factory at sql-land means including its header there and linking pybind to the main lib, which I guess is a no-no
Or am I missing something?

Reply:
Note that test_arrow, is just a copy of the actual structure in arrow_array_stream.hpp to allow tests in the main lib.

Reply:
Ah, got it. I didn't notice that the main arrow conversion function lives inside the main library. I think that's fine then.